### DISCUSSION TITLE
How can I access `NSWorkspace` without `MonoMac.NetStandard` in AvaloniaUI?

### DISCUSSION THREAD

**Discussion by Irasil** [AvaloniaUI/Avalonia#13455] [Q&A] [answered]
I'm completely new to AvaloniaUI, and I'm a programming apprentice, so please forgive me if this is a basic question.

I want to develop a cross-platform app with Avalonia that monitors processes. For Windows, I'm using `ManagementEventWatcher(query)`, and it's working fine. On macOS, I need to access `NSWorkspace.SharedWorkspace.NotificationCenter` to achieve what I need. I managed to do this using the NuGet package `MonoMac.NetStandard`, but my supervisor doesn't want to use it because it seems it's no longer actively developed.

Are there other ways to access `NSWorkspace` from AvaloniaUI? I've heard there's an option to target .Net6.0-macos, but that would make it not work on Windows anymore, or do I need to create two separate projects? If it's about using .Net6.0-macos, can someone provide a detailed example of how to implement it?

Here's a code snippet I'm trying to get to work.

```csharp
 public void WatchProcesses()
 {
     var workspace = NSWorkspace.SharedWorkspace;
     var notificationCenter = NSWorkspace.SharedWorkspace.NotificationCenter;

     var appLaunchObserver = notificationCenter.AddObserver(NSWorkspace.DidLaunchApplicationNotification, (notification) =>
     {
         var key = new NSString("NSApplicationName");
         var appName = notification.UserInfo[key];
         Console.WriteLine(appName);
     });
 }

```

I appreciate any help in advance!

**Comment by thevortexcloud** [comment]
If you need to use a platform specific TFM the general approach is to make a shared project all platforms can access and then have a different executable per platform TFM. This is required because you can't have a generic TFM depend on a platform specific ones (for some reason I never understood as there is no technical reason why most of them can't compile from what I can tell, and then do runtime checks yourself). You can technically get around that if you are willing to implement the native interop yourself using things like `DllImport` but it can be a lot of work and fairly complicated. But it does simplify your project dependencies a lot. In any case, the Avalonia cross platform template that ships with Avalonia uses different projects per platform and uses platform specific TFMs which may do what you want.

In any case the official replacement for MonoMac is apparently Xamarin MaciOS https://github.com/xamarin/xamarin-macios

Xamarin (and Mono) is owned by Microsoft so it should be fairly well supported for a long time. Although I think that just might be the Mac SDK that the Mac TFM uses. I don't ever program for Macs so I am somewhat useless on the specifics of what happens when you use the Mac TFM.

**Reply by Irasil** [reply]
Thank you very much for the quick response!

I did try using `DllImport` but failed miserably.

```csharp

[DllImport("/System/Library/Frameworks/Foundation.framework/Foundation")]
public static extern IntPtr NSNotificationCenter_defaultCenter();
```
This resulted in the error "Unable to find an entry Point named NSNotificationCenter_defaultCenter() in the shared library System/Library/Frameworks/Foundation.framework/Foundation."

Somewhere, I read that methods must be offered as P/Invoke, and if they are not, you cannot call them via P/Invoke. But as I mentioned, this is all new to me, and I don't quite have the understanding to be sure. Still, I've put this approach aside.

I've also come across Xamarin MaciOS, but I don't really understand how to use it, as it's not a NuGet package. The installation instructions for Windows are not very clear. It seems like it's installed when you install Visual Studio for Mac. 

So my next attempt will be to create my own project for MacOS-specific methods.

In the image here, you can see what I can choose from when I use the Mac:

![Screenshot 2023-11-01 152242](https://github.com/AvaloniaUI/Avalonia/assets/91846030/52639a79-2420-4a83-be89-31a2525dfd42)

When I choose "macOS Application," I can select .NET 6-8 as the framework, and when I choose "Cocoa App," I only have .NET 4.x to choose from. Which one should I use? 

I'm also not entirely sure about the long-term support for Xamarin, as Microsoft is pushing for MAUI. Support for Xamarin is set to be discontinued next May, or does that only apply to Xamarin.Forms and not Xamarin.Mac? [Xamarin Support Policy](https://dotnet.microsoft.com/en-us/platform/support/policy/xamarin#:~:text=Xamarin%20support%20will%20end%20on,about%20upgrading%20Xamarin%20projects%20to%20.)

**Reply by thevortexcloud** [reply]
> This resulted in the error "Unable to find an entry Point named NSNotificationCenter_defaultCenter() in the shared library System/Library/Frameworks/Foundation.framework/Foundation."

I know nothing about writing native code for Mac. But method names are case sensitive for `DllImport`, and you generally point `DllImport` at the system library itself. EG `libc.so` or `libc`. Also not all native libraries can be used. They need to be compiled in a way that makes them compatible with C. As C is the IDL. This means things like libraries written in Rust can't be used by C# by default. 

>I'm also not entirely sure about the long-term support for Xamarin, as Microsoft is pushing for MAUI. Support for Xamarin is set to be discontinued next May, or does that only apply to Xamarin.Forms and not Xamarin.Mac? [Xamarin Support Policy](https://dotnet.microsoft.com/en-us/platform/support/policy/xamarin#:~:text=Xamarin%20support%20will%20end%20on,about%20upgrading%20Xamarin%20projects%20to%20.)

Apparently it's just included in the Mac TFM now. Which is not going anywhere.

https://learn.microsoft.com/en-us/dotnet/standard/frameworks


>When I choose "macOS Application," I can select .NET 6-8 as the framework, and when I choose "Cocoa App," I only have .NET 4.x to choose from. Which one should I use?

If you are making a new app, you absolutely should avoid targeting .NET Framework. The only way to run that on non Windows platforms is via Mono. Mono _generally_ works, But it has a lot of missing features/bugs too. It's really only supported these days on android/iOS. Not to mention you are stuck with what ever features .NET 4.8 shipped with instead of newer versions of .NET. That said Mono can do things official modern .NET can't do either on Mac (eg Win Forms)

If you want to try Avalonia specifically, try running `dotnet new avalonia.xplat` in a terminal which will generate a solution which will work on Windows, Mac, Linux, web, android and iOS using the various platform specific TFMs as needed. 

That said if you don't care about making your app work cross platform you can just target `net6.0-macos` everywhere.

**Reply by Irasil** [reply]
Thank you for your help, I appreciate that!

> I know nothing about writing native code for Mac. But method names are case sensitive for DllImport, and you generally point DllImport at the system library itself. EG libc.so or libc. Also not all native libraries can be used. They need to be compiled in a way that makes them compatible with C. As C is the IDL. This means things like libraries written in Rust can't be used by C# by default.

If I can't find another way, I'll have to deal with `DllImport` again.

> Apparently it's just included in the Mac TFM now. Which is not going anywhere.
https://learn.microsoft.com/en-us/dotnet/standard/frameworks

I tried to reference two projects with different TFM yesterday, but it didn't work. Through your link, I came across `<TargetFrameworks>` in plural, and I will try that now.

> If you are making a new app, you absolutely should avoid targeting .NET Framework. The only way to run that on non Windows platforms is via Mono. Mono generally works, But it has a lot of missing features/bugs too. It's really only supported these days on android/iOS. Not to mention you are stuck with what ever features .NET 4.8 shipped with instead of newer versions of .NET. That said Mono can do things official modern .NET can't do either on Mac (eg Win Forms)

I thought so, and I tried it with .NET 6.0 yesterday. I was just confused because I originally wanted to try Xamarin.Mac and thought I had to choose Xamarin.

> If you want to try Avalonia specifically, try running dotnet new avalonia.xplat in a terminal which will generate a solution which will work on Windows, Mac, Linux, web, android and iOS using the various platform specific TFMs as needed.

I briefly looked at the Avalonia Cross-Platform application, and I don't think it's a viable option. I need to develop a new app that runs on Windows, MacOS, and Linux. I require platform-specific system APIs for each platform since I need to react to starting processes without actively polling them at intervals, for example.

As far as I can see, Avalonia Cross-Platform application supports all the platforms, but the desktop variants are grouped together, and I don't think that would work. I'll come back to this later if I can't find any other options.

> That said if you don't care about making your app work cross platform you can just target net6.0-macos everywhere.

"Unfortunately," the app needs to be cross-platform compatible.

**Reply by thevortexcloud** [reply]
> I tried to reference two projects with different TFM yesterday, but it didn't work. Through your link, I came across <TargetFrameworks> in plural, and I will try that now.

Multitargeting will run into most of the same problems. The only differences are you can use a bunch of preprocessor directives to change what code gets used at compile time (EG don't use Mac specific code on Windows) and it will build a different assembly for each target. It could work for this. But preprocessors can quickly get out of hand as you wont know about build errors for targets you are not building for. So it's fairly easy to break things without realising it. Which is why I personally prefer not to use them. But if you don't mind that or feel like writing a full unit test suite to reduce the chance of missing platform specific compile time errors, then it's a viable option. That said if you are not using any platform specific code but are using the platform TFM as another target, it will probably get around the build error anyway.

If you really do need to use the MacOS TFM, my advice would be to make a shared project with as much common functionality as possible, then define a bunch of interfaces for platform specific stuff. Then implement that in the platform specific code, and pass the interface instance back to the common code where your main application logic resides (this is where good design patterns are helpful, EG using dependency injection). But this can also get fairly messy and means you still need multiple executables compared to the one you would need if you implemented all the interop yourself. You will also probably end up duplicating some code in all platforms.

Alternatively, but also far more annoying to manage you could probably use a plugin architecture. Which means you load your DLL at runtime and then try to load some class from it. It means you only need a single executable but it also means you need to deal with non compile time checks. EG it's possible for the plugin to be compiled against an assembly that has a method, and then that method is removed. If the plugin then tries to use that method it will throw an error.

> I briefly looked at the Avalonia Cross-Platform application, and I don't think it's a viable option. I need to develop a new app that runs on Windows, MacOS, and Linux. I require platform-specific system APIs for each platform since I need to react to starting processes without actively polling them at intervals, for example.

It does that already. It has a different executable per platform TFM, and you can easily change what the target is or add another. You just need to architect in a way you can share logic with all platforms (like what I mentioned above).

**Reply by thevortexcloud** [reply]
> I need to react to starting processes without actively polling them at intervals, for example.

Oh also, you can technically do this on Linux by just setting up a `FileSystemWatcher` and pointing it at `/proc/`. It's internally implemented using [`INotify`](https://man7.org/linux/man-pages/man7/inotify.7.html) on Linux. Which means you can do it in pure crossplatform compatible C# code (it obviously wont actually work on other platforms).

**Reply by Irasil** [reply]
I attempted multitargeting, but the approach I wanted didn't work, and as you described, there are many pitfalls.

> If you really do need to use the MacOS TFM, my advice would be to make a shared project with as much common functionality as possible, then define a bunch of interfaces for platform specific stuff. Then implement that in the platform specific code, and pass the interface instance back to the common code where your main application logic resides (this is where good design patterns are helpful, EG using dependency injection). But this can also get fairly messy and means you still need multiple executables compared to the one you would need if you implemented all the interop yourself. You will also probably end up duplicating some code in all platforms.

I tried this approach, but I feel like I did it the wrong way. I have an Avalonia project (TFM .net6.0) that includes platform-specific services. Then, I created a `macOS Class Library project (.net6.0-macos)` and added it as a reference to the Avalonia project. However, this doesn't work because a `.net6.0` project cannot reference a `.net6.0-macos` project. It works the other way around; I can add the Avalonia project as a reference to the macOS Class Library project.
Then, in the Avalonia project, I created an interface and implemented it in the macOS Class Library project. But how do I get that implementation back into the Avalonia project since I don't have a reference to the macOS Class Library project?

Here's an image of my solution:
![Screenshot 2023-11-02 132515](https://github.com/AvaloniaUI/Avalonia/assets/91846030/d19cee74-c8a6-47c8-b406-1f0489acd243)

Test (.net6.0-macos) references AvaloniaTest
Class1: ITest

Have I understood this correctly, or do I need to do it the other way around? Should I create a main project for each platform, each of which references the shared project?

> Alternatively, but also far more annoying to manage you could probably use a plugin architecture. Which means you load your DLL at runtime and then try to load some class from it. It means you only need a single executable but it also means you need to deal with non compile time checks. EG it's possible for the plugin to be compiled against an assembly that has a method, and then that method is removed. If the plugin then tries to use that method it will throw an error.

As I understand it, this is a bit beyond my current knowledge level.

> Oh also, you can technically do this on Linux by just setting up a FileSystemWatcher and pointing it at /proc/. It's internally implemented using [INotify](https://man7.org/linux/man-pages/man7/inotify.7.html) on Linux. Which means you can do it in pure crossplatform compatible C# code (it obviously wont actually work on other platforms).

Thank you for the advice; I'll keep it in mind once I get macOS up and running.
I can't express how grateful I am for your time!

**Reply by thevortexcloud** [reply]
> Have I understood this correctly, or do I need to do it the other way around? Should I create a main project for each platform, each of which references the shared project?

You should set up your DI container in each platform specific project. Since you can't do it in the shared project. This should work fine since it will be setup on boot anyway. You can however make some helper/extension methods in the shared project to register common services. Since I am fairly lazy, I typically make one to automagically register my views/viewmodels for injection using reflection. But that's not suitable for all projects/circumstances (as you would likely not want to register all views/viewmodels as many may need to be created manually). You do however need to remember to add all the required services for all platforms. As you wont get a compiler error for it.

> As I understand it, this is a bit beyond my current knowledge level.

It's actually far simpler than it sounds to implement it. Way simpler than using `DllImport`. You really just need [this](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.assembly.load?view=net-7.0) method. to load DLLs that are not referenced at compile time. Then you check the exported types from the loaded assembly for what ever class you want to instantiate, and then you can use [`Activator.CreateInstance`](https://learn.microsoft.com/en-us/dotnet/api/system.activator?view=net-7.0) to create an instance of it. Or you can add the type/instance to your DI container so what ever services it exposes can be used with DI (EG searching for all implementations of a specific interface). I typically either load plugins from a specific directory and also require plugin assemblies to implement some sort of interface declared in a shared project to indicate it's a plugin project. Which makes finding plugins assemblies when searching assembly easier. EG you could do something like

```csharp
interface IPlugin {
string PluginName {get;}
void RegisterServices(ServiceCollection services);
}
```
So when you create the instance of the class, you just call the register services method and pass it the collection to register services.

But it's certainly not an optimal way to develop something and maintenance can be a pain. It works really well though if you need to be able to run other people's code within your app and they can't modify your source code. But at that point you probably also need to promise some sort of stable API to avoid breaking their code when you change yours.

tl;dr Reflection is very powerful. But can also be very painful.

Personally though, I prefer what ever approach can give me the most amount of compile time checks.

>I can't express how grateful I am for your time!

Happy to help.

**Reply by Irasil** [reply]
> You should set up your DI container in each platform specific project. Since you can't do it in the shared project. This should work fine since it will be setup on boot anyway. You can however make some helper/extension methods in the shared project to register common services. Since I am fairly lazy, I typically make one to automagically register my views/viewmodels for injection using reflection. But that's not suitable for all projects/circumstances (as you would likely not want to register all views/viewmodels as many may need to be created manually). You do however need to remember to add all the required services for all platforms. As you wont get a compiler error for it.

I apologize for my misunderstanding, but could you elaborate on this point further? Perhaps even provide an example. I'm sorry for asking so much, but I didn't quite grasp it.

I originally wanted to separate platform-specific methods because they are not present in Avalonia, but still maintain a main project for all platforms that defines the entire app's lifecycle.

If I understand you correctly, you're suggesting that I should create a separate app with its own lifecycle for each platform, and the views would be shared from a shared project. What I don't understand is how it's decided which project to run. In my initial attempt, I simply checked on which platform I was running in App.axaml.cs and then loaded the corresponding service.

Am I correct in understanding that if I implement it as you've suggested, I will have to manually choose which project to start, much like in an Avalonia cross-platform app? And when I compile it, will it generate three different applications?

**Reply by thevortexcloud** [reply, marked as answer]
> I apologize for my misunderstanding, but could you elaborate on this point further? Perhaps even provide an example. I'm sorry for asking so much, but I didn't quite grasp it.

A somewhat simple example using Microsoft's DI would be:

```csharp
public static class ServiceCollectionExtensions {
    public static void AddCommonServices(this IServiceCollection collection) {
        collection.AddTransient<MainViewModel>();
    }
}
```
```csharp
public static class ServiceManager {
    public static IServiceProvider Locator { get; private set; }

    public static void InitialiseLocatorFromCollection(IServiceCollection collection) {
        Locator = collection.BuildServiceProvider();
    }
}
```
```csharp
    [STAThread]
    public static void Main(string[] args) {
        var collection = new ServiceCollection();
        collection.AddCommonServices();
        //Register platform specific services here with the collection
        
        ServiceManager.InitialiseLocatorFromCollection(collection);
        
        BuildAvaloniaApp()
            .StartWithClassicDesktopLifetime(args);
    }

    // Avalonia configuration, don't remove; also used by visual designer.
    public static AppBuilder BuildAvaloniaApp()
        => AppBuilder.Configure<App>()
            .UsePlatformDetect()
            .WithInterFont()
            .LogToTrace()
            .UseReactiveUI();
}
```

```csharp
public partial class App : Application {
    public override void Initialize() {
        AvaloniaXamlLoader.Load(this);
    }

    public override void OnFrameworkInitializationCompleted() {
        var vm = ServiceManager.Locator.GetRequiredService<MainViewModel>();
        if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop) {
            desktop.MainWindow = new MainWindow {
                DataContext = vm,
            };
        }
        else if (ApplicationLifetime is ISingleViewApplicationLifetime singleViewPlatform) {
            singleViewPlatform.MainView = new MainView {
                DataContext = vm
            };
        }

        base.OnFrameworkInitializationCompleted();
    }
}
```

That said to really take advantage of this approach you also need to write some infrastructure code (which you will only need to do once). For views you can make a simple [ViewLocator](https://docs.avaloniaui.net/docs/next/concepts/view-locator) which will automatically resolve the view from the `Locator`. For viewmodels you can make some sort of service that does similar things and automatically assigns it to a view. I typically make some sort of window manager which is injected into to all top level views which handles resolving and displaying things like windows for me without directly coupling it to the locator or manually creating any of the views/most of the view models. You should also avoid using the `Locator` whenever possible for many reasons (the main ones being it creates a lot of coupling and makes unit tests way harder to write). But it can't always be avoided. But you absolutely should not be using it inside a view model since those will have injected services.

I should also note Avalonia actually has a built in DI framework called [Splat](https://github.com/reactiveui/splat). It also has a locator. It's not as flexible though as the Microsoft one but it works fine for most things. You can technically use Microsoft's DI with Splat as well, but it requires additional setup to use with Avalonia.

> Am I correct in understanding that if I implement it as you've suggested, I will have to manually choose which project to start, much like in an Avalonia cross-platform app? And when I compile it, will it generate three different applications?

This is correct. It's what Microsoft wants people to do. I also agree, it feels clunky and not very nice in most instances. As mentioned previously there are ways around it (at least on platforms that can use the same binaries, such as desktops). But it's a lot more work in many cases. But at least there is no Linux specific target. So all your Linux specific code can be "crossplatform".

**Reply by Irasil** [reply]
Thank you very much, you saved me. This approach seems to work for the moment. Now I have a foundation on which I can build. I sincerely thank you for your time and efforts.

Greetings,
Irasil

**Reply by thevortexcloud** [reply]
You are welcome! If you need further help with something else, let me know or open a new issue.

**Reply by Irasil** [reply]
I was able to get it running with your instructions so that I can use platform-specific API's. However, now I have the problem that the previewer no longer works.
I had to modify your instructions a bit to make it work the way I want.

I adopted ServiceManager.cs and ServiceCollectionExtension.cs from you.

My Program.cs

```csharp
[STAThread]
public static void Main(string[] args)
{
    var collection = new ServiceCollection();
    collection.AddCommonServices();
    
    collection.AddSingleton<IPlatformSpecificServices>(new Platform());

    ServiceManager.InitialiseLocatorFromCollection(collection);

    BuildAvaloniaApp()
        .StartWithClassicDesktopLifetime(args);
}

// Avalonia configuration, don't remove; also used by visual designer.
public static AppBuilder BuildAvaloniaApp()
    => AppBuilder.Configure<App>()
        .UsePlatformDetect()
        .WithInterFont()
        .LogToTrace();
```

My App.axaml.cs

```csharp

 public Window MainWindowInstance { get; set; }
 public override async void OnFrameworkInitializationCompleted()
 {
     var platformSpecificServices = ServiceManager.Locator.GetRequiredService<IPlatformSpecificServices>();
     var vm = ServiceManager.Locator.GetRequiredService<MainViewModel>();

     if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime)
     {
         MainWindowInstance = new MainWindow(platformSpecificServices)
        {
             DataContext = vm
        };

         MainWindowInstance.Closing += (sender, e) =>
         {
             e.Cancel = true;
             MainWindowInstance.Hide();
         };
     }
     else if (ApplicationLifetime is ISingleViewApplicationLifetime singleViewPlatform)
     {
         singleViewPlatform.MainView = new MainView();
     }
 }

 public void ShowWindow_Click(object sender, EventArgs e)
 {
     if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
     {
         desktop.MainWindow = MainWindowInstance;
         desktop.MainWindow.Show();
     }
 }
```

Here is the error I get from the previewer:

```txt
09:16:28.900 [Verbose] 36972 Started PreviewerProcess.ConnectionInitializedAsync()
09:16:28.901 [Information] 36972 Connection initialized
09:16:28.901 [Debug] 36972 => Sending ClientSupportedPixelFormatsMessage { Formats: [Bgra8888, Rgba8888] }
09:16:28.902 [Debug] 36972 => Sending ClientRenderInfoMessage { DpiX: 96, DpiY: 96 }
09:16:28.902 [Verbose] 36972 Finished PreviewerProcess.ConnectionInitializedAsync()
09:16:28.903 [Verbose] 36972 Finished PreviewerProcess.StartAsync()
09:16:28.903 [Debug] 36972 => Sending UpdateXamlMessage { Xaml: "<UserControl xmlns=\"https://git…", AssemblyPath: "C:\Users\test\repositorie…", XamlFileProjectPath: null }
09:16:28.904 [Verbose]  Finished AvaloniaDesigner.StartProcessAsync()
09:16:28.906 [Debug] 36972 <= "Obtaining AppBuilder instance from Test.Avalonia.Desktop.Program"
09:16:28.935 [Debug] 36972 <= "Initializing application in design mode"
09:16:29.043 [Debug] 36972 <= "Sending StartDesignerSessionMessage"
09:16:29.047 [Verbose] 36972 Started PreviewerProcess.OnMessageAsync()
09:16:29.048 [Debug] 36972 <= StartDesignerSessionMessage { SessionId: "f7e259fc-f622-4c44-9a0c-4ada9c6…" }
09:16:29.048 [Verbose] 36972 Finished PreviewerProcess.OnMessageAsync()
09:16:29.055 [Error] 36972 <= "Unhandled exception. System.ArgumentNullException: Value cannot be null. (Parameter 'provider')"
09:16:29.056 [Error] 36972 <= "   at System.ThrowHelper.Throw(String paramName)"
09:16:29.056 [Error] 36972 <= "   at System.ThrowHelper.ThrowIfNull(Object argument, String paramName)"
09:16:29.056 [Error] 36972 <= "   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)"
09:16:29.058 [Error] 36972 <= "   at Test.Avalonia.App.OnFrameworkInitializationCompleted() in C:\Users\test\App.axaml.cs:line 30"
09:16:29.059 [Error] 36972 <= "   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)"
09:16:29.059 [Error] 36972 <= "   at Avalonia.Threading.SendOrPostCallbackDispatcherOperation.InvokeCore()"
09:16:29.059 [Error] 36972 <= "   at Avalonia.Threading.DispatcherOperation.Execute()"
09:16:29.060 [Error] 36972 <= "   at Avalonia.Threading.Dispatcher.ExecuteJob(DispatcherOperation job)"
09:16:29.060 [Error] 36972 <= "   at Avalonia.Threading.Dispatcher.ExecuteJobsCore(Boolean fromExplicitBackgroundProcessingCallback)"
09:16:29.060 [Error] 36972 <= "   at Avalonia.Threading.Dispatcher.Signaled()"
09:16:29.060 [Error] 36972 <= "   at Avalonia.Controls.Platform.ManagedDispatcherImpl.RunLoop(CancellationToken token)"
09:16:29.061 [Error] 36972 <= "   at Avalonia.Threading.DispatcherFrame.Run(IControlledDispatcherImpl impl)"
09:16:29.061 [Error] 36972 <= "   at Avalonia.Threading.Dispatcher.PushFrame(DispatcherFrame frame)"
09:16:29.062 [Error] 36972 <= "   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken)"
09:16:29.062 [Error] 36972 <= "   at Avalonia.DesignerSupport.Remote.RemoteDesignerEntryPoint.Main(String[] cmdline)"
09:16:29.062 [Error] 36972 <= "   at Avalonia.Designer.HostApp.Program.Main(String[] args)"
09:16:32.013 [Information]  Process exited
09:16:32.013 [Verbose]  Started PreviewerProcess.Stop()
09:16:32.014 [Information]  Stopping previewer process
09:16:32.014 [Verbose]  Finished PreviewerProcess.Stop()
09:16:32.013 [Error]  Connection error
```

I am very grateful for any help!

**Reply by maxkatz6** [reply]
Previewer doesn't run Main method, it runs BuildAvaloniaApp directly. Alternatively, you can setup your DI in the App.axaml.cs file. Or in the `.AfterSetup()` extension method of AppBuilder.

**Reply by thevortexcloud** [reply]
From memory the reason I said to do it in the main method is the app.axaml is the shared project which does not have access to all the platform specific services.  But I have never seen anything use the `AfterSetup` before. So that's also good to know!

**Reply by Irasil** [reply]
I apologize for asking repeatedly, but unfortunately, I don't understand it.

If I understood correctly, I need to do it in .AfterSetup().

But how exactly can I insert the DI there?

Could you give me an example?

As I said, the support here is incredibly good, I've never experienced anything like it, many thanks!

**Reply by Irasil** [reply]
I think I've figured it out:

Here's my new Program.cs

```csharp
  public static AppBuilder BuildAvaloniaApp()
      => AppBuilder.Configure<App>()
          .UsePlatformDetect()
          .WithInterFont()
          .LogToTrace()
          .AfterSetup(SetupDependencyInjection);

  private static void SetupDependencyInjection(AppBuilder builder)
  {
      var collection = new ServiceCollection();
      collection.AddCommonServices();
      collection.AddSingleton<IPlatformSpecificServices>(new Platform());

      ServiceManager.InitialiseLocatorFromCollection(collection);
  }
```

After that, I had to adjust MainViewModel.cs, so I added a constructor for the designer.

```csharp
 private readonly IPlatformSpecificServices _services;

 public MainViewModel(IPlatformSpecificServices services)
 {
     _services = services;
 }

 public MainViewModel()
 {
     if (Design.IsDesignMode)
     {
     }
 }
```

Is that correct?
I thank you all very much.

**Reply by thevortexcloud** [reply]
You can make a custom markup extension to resolve services at design time: https://github.com/AvaloniaUI/Avalonia/discussions/13743#discussioncomment-7686499

Which means you don't need the second constructor. Adding a parameterless constructor may break your DI in subtle ways.

**Reply by ivoryguard** [reply]
@thevortexcloud - You saved me a lot of time! 

I succeeded in implementing dependency injection for platform specific features based on my .NET experiences (Xamarin.Forms, Blazor wasm an so on) and your sample codes.

The documentation for Avalonia's cross-platform implementation is too lacking. Implementing platform specific features is very essential topic. 

In Xamarin.Forms, DI is explained in detail in both the official documentation and samples, but I think Avalonia may leave these explanations to other packages like DI and so on.

These make it difficult for beginners to dive into Avalonia.

**Reply by thevortexcloud** [reply]
> The documentation for Avalonia's cross-platform implementation is too lacking. Implementing platform specific features is very essential topic.

> In Xamarin.Forms, DI is explained in detail in both the official documentation and samples, but I think Avalonia may leave these explanations to other packages like DI and so on.

There is now an official Avalonia DI guide which I helped write.

https://docs.avaloniaui.net/docs/guides/implementation-guides/how-to-implement-dependency-injection

However it was decided to keep it simple and worry about how to implement it properly for cross platform scenarios later. So it only tells you how to implement it for a single platform (IE all desktop platforms).

https://github.com/AvaloniaUI/avalonia-docs/pull/359

I imagine a PR would be welcome adding information on how to do DI across projects like this. Avalonia itself though is primarily a UI framework. All the UI functionality is abstracted and can (generally speaking) be used entirely from a single project. You really only need to do this if you need access to non UI related platform specific functionality. Which most people don't. Xamarin on the other hand was a complete framework for creating apps across platforms, not just GUIs. Also Avalonia itself does not support DI out of the box, you need to add it yourself. How you do it is really up to you,

I should note as was mentioned in other places in this discussion, you should not create the DI container in the main method.

On a related note: I still really wish Microsoft would drop the TFM restrictions and let you do runtime/custom compile time checks for accessing the TFM specific APIs in library projects. It would make structuring all of this significantly simpler. If you add the exact same functionality yourself in via a `DllImport` you wont hit the same problems and everything will happily compile in a single project.

**Reply by ivoryguard** [reply]
I am currently working on porting my classic Windows WPF apps to mobile. 

There are several frameworks I can choose. 
I had a terrible experience with Xamarin.Forms, so I have no intention of using MAUI. 

Then, the remaining alternatives are Uno and Avalonia. (OpenSilver is for web site not application basically,)
Uno's rendering method has similarities to MAUI, which is why I think Avalonia, which uses a rendering method similar to Flutter
So, I think that Avalonia is the best option fto port my classic WPF app to mobile.

I already know that Avalonia is a UI Framework. 

In the past, Avalonia may be desktop-centric, but I believe that starting from Avalonia 11, there will be a significant increase in developers aiming to use it for mobile-first cross-platform app development.

Therefore, I hope that guides and sample sources to develop mobile-first Avalonia cross-platform apps will improve, even if they are not directly related topics to Avalonia itself.

Many classic WPF apps do not use DI and ReactiveUI which are described in Avalonia doc and samples.
In my case, my classic WPF apps for .NET Framework 4.0~4.7.2 use MVVM without any external MVVM framework, and do not apply DI . So I prefer Community MVVM Toolkit instead of ReactiveUI in order to decease code changes for porting :)

This post is very helpful and I appreciate your efforts.

**Reply by thevortexcloud** [reply]
> I had a terrible experience with Xamarin.Forms, so I have no intention of using MAUI.

That's fair. The general consensus I last heard seemed to be that MAUI was still broken/missing features and therefore unusable. Although MAUI for me personally is completely unusable due to Microsoft consistently deciding that nobody could possibly ever need a GUI on a Linux system ([despite the very large amount of people providing business cases to the contrary](https://github.com/dotnet/maui/discussions/339), and Microsoft refusing to acknowledge that Linux is used everywhere outside the desktop including in embedded devices).

>  but I believe that starting from Avalonia 11, there will be a significant increase in developers aiming to use it for mobile-first cross-platform app development.

There has certainly been a lot of fixes to make the experience better. It's still not as good as it could be though. Although there are some third party control libraries trying to make it feel more mobile native. EG the [material theme ](https://github.com/AvaloniaCommunity/Material.Avalonia) and the[ shell project.](https://github.com/AvaloniaInside/Shell)

The other problem with mobile is that a lot of third party controls and things were just never designed to be run on mobile devices. They will run great on desktops but then get ~3fps when run on a mid range phone from ~4 years ago. But that's an optimisation problem.

> This post is very helpful and I appreciate your efforts.

You are welcome!

**Comment by thevortexcloud** [comment]
> now I have the problem that the previewer no longer works.

In my experience, it breaks as soon as you do anything mildly complicated and requires a lot of effort to work around. Which is why I don't bother using it most of the time. However, in this case I am guessing it's failing because it does not know how to construct your view. If you add a parameterless constructor it will probably get further.

It might however work with a custom view locator that resolves your window from your container. I can't remember if the previewer will use those for a top level view (I think it should?).

https://docs.avaloniaui.net/docs/concepts/view-locator

If it does, you just need to make sure your view is set up for DI. I would actually do that anyway it saves you from manually instantiating your view,

Otherwise you should probably try to move as much as you can into your VM.

**Reply by maxkatz6** [reply]
> In my experience, it breaks as soon as you do anything mildly complicated and requires a lot of effort to work around

In this case reason is pretty simple, as Main method isn't getting executed.

**Reply by thevortexcloud** [reply]
That's genuinely good to know!
